orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py

```diff
@@ -111,6 +119,8 @@ def _export_pt_1_10(g, n, *args, **kwargs):
                     # All other inputs are accessed via "pointers".
                     input_pointer_scalar_positions.append(i)
                     input_pointer_scalars.append(id(arg))
+
+                    NONTENSOR_OBJECT_POINTER_STORE[id(arg)] = arg
```

Comment:
This can be a workaround. For a sustainable fix, do you feel we can increase its reference count here and decrease reference count in PythonOp's dtor? Or we should provie an API to release  objects in NONTENSOR_OBJECT_POINTER_STORE? For example,
```python
# training loop
for ....
  ort.start_cache_object_pointers()
  loss = model(x)
  loss.backward()
 optimizer.step()
 ...
  ort.release_pointers_cached_in_this_iteration()
```
You can use `with` as well, e.g.,
```python
for ....
  with object_pointer_pool():
    loss = model(x)
    loss.backward()
    optimizer.step()
 ...
```

Reply:
Imagine pointers and other non-tensor constant arguments are "constant", so ideally once model is exported, the pointers should also be valid unless the model ends its lifetime. This is how I understand their relation. 

One usage pattern looks like this to end the scope before compute runs:

idx=0
while idx < 12:
  group=create_process_group()
  m = prepare_torch_model(group)
  m = ORTModule(m)
  loss = torch.sum(m(input))
  loss.backward()

when idx=0 loop complete it should be fine.
when idx=1, group is overridden, if the idx=0's cuda compute (PythonOp) did not complete, it will use a release buffer, which is unexpected. 

Add more scope in user script did not helps for this case.

Reply:
Let PythonOp kernels maintain the ref make some senses, there is one trick issue there. Imagine PythonOp kernel creation, we increase the ref cnt by 1, in its destructor, we decrease the ref cnt by 1.  If Python program exit, but C++ did not complete destructors for those kernels, An invalid pythonobject will be released by kernel's destructors, a segement fault will happen then, unless we have some ways to trigger those releasing proactively before python program exit (using atexit).

Reply:
Given that this arg is a non tensor type, can we make a deepcopy of it and save it to the pythonop attribute? Then the ownership of the copy will be with the pythonop.

Reply:
> Imagine pointers and other non-tensor constant arguments are "constant", so ideally once model is exported, the pointers should also be valid unless the model ends its lifetime. This is how I understand their relation.
> 
> One usage pattern looks like this to end the scope before compute runs:
> 
> idx=0 while idx < 12: group=create_process_group() m = prepare_torch_model(group) m = ORTModule(m) loss = torch.sum(m(input)) loss.backward()
> 
> when idx=0 loop complete it should be fine. when idx=1, group is overridden, if the idx=0's cuda compute (PythonOp) did not complete, it will use a release buffer, which is unexpected.
> 
> Add more scope in user script did not helps for this case.

I think PyTorch has the same problem with this example, so we can ignore it. PyTorch users should write PyTorch-compatible code.

Reply:
> Let PythonOp kernels maintain the ref make some senses, there is one trick issue there. Imagine PythonOp kernel creation, we increase the ref cnt by 1, in its destructor, we decrease the ref cnt by 1. If Python program exit, but C++ did not complete destructors for those kernels, An invalid pythonobject will be released by kernel's destructors, a segement fault will happen then, unless we have some ways to trigger those releasing proactively before python program exit (using atexit).

If use `atexit` solves the problem, it's fine, right?

Reply:
> Given that this arg is a non tensor type, can we make a deepcopy of it and save it to the pythonop attribute? Then the ownership of the copy will be with the pythonop.

Oh, cloning a torch process group might be tricky, I am not sure whether the object is self-contained, or clone-able. Maybe we can revisit if we need such kind of fix later.

Reply:
Talked with @wschin, the conclusion is, we do this way without introducing more complex in C++ backend side. A python interface is also can be used for potential usage if customers find those object references are not released in time (which should not be a common usage case).